[PKG-2813] Rebuild pre-commit 3.2.1 to fix win64

skupr-anaconda opened the pull request:
Changelog: https://github.com/pre-commit/pre-commit/blob/v3.2.1/CHANGELOG.md
License: https://github.com/pre-commit/pre-commit/blob/v3.2.1/LICENSE
Requirements: https://github.com/pre-commit/pre-commit/blob/v3.2.1/setup.cfg

Actions:
1. Increase the build number to `1`
2. Add `--no-build-isolation` to build `scripts`
3. Rename package to `pre-commit` because `conda-build` incorrectly creates `entry_ponits` on `win64`. I guess it's a **conda-build** bug. I left a comment on why not rename it again the next time
4. Move tests from the output `pre-commit` to the top level of the recipe
5. Move test commands from test scripts to the `test/commands` section

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:34: host_section_needs_exact_pinnings: output "pre-commit": Packages in host must have exact version pinnings, except python build tools.
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
ERROR: clone/recipe/meta.yaml:37: host_section_needs_exact_pinnings: output "pre-commit": Packages in host must have exact version pinnings, except python build tools.
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:25: uses_setup_py: `python setup.py install` is deprecated.
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

JeanChristopheMorinPerso commented on recipe/meta.yaml:
```diff
@@ -23,8 +22,8 @@ requirements:
-    script: build_base.bat  # [win]
-    script: build_base.sh  # [not win]
+    build:
+      script: python setup.py install --single-version-externally-managed --record=record.txt
```
I'm not sold on this. I'd prefer if we find the real problem then doing this.

skupr-anaconda replied:
My suggestion is that the issue is caused by `pip` or `virtualenv` (or incompatible versions) but I haven't any proves yet

cbouss replied:
cf pre-commit is a noarch package. It doesn't include file pre-commit.exe, unlike our package.

JeanChristopheMorinPerso replied:
`pre_commit` is noarch but `pre-commit` has a per python build.

JeanChristopheMorinPerso replied:
Ah wait, by `cf` you mean conda-forge?

cbouss replied:
yes, if you extract both the difference is there

skupr-anaconda replied:
If I make a noarch package then the error will be the same

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
Traceback (most recent call last):
  File "/opt/conda/envs/linter/bin/conda-lint", line 11, in <module>
    sys.exit(main())
             ^^^^^^
  File "/opt/conda/envs/linter/lib/python3.11/site-packages/anaconda_linter/run.py", line 114, in main
    result = linter.lint(
             ^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.11/site-packages/anaconda_linter/lint/__init__.py", line 626, in lint
    msgs = self.lint_file(
           ^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.11/site-packages/anaconda_linter/lint/__init__.py", line 730, in lint_file
    self.lint_recipe(
  File "/opt/conda/envs/linter/lib/python3.11/site-packages/anaconda_linter/lint/__init__.py", line 784, in lint_recipe
    res = self.check_instances[check].run(recipe, fix)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/conda/envs/linter/lib/python3.11/site-packages/anaconda_linter/lint/__init__.py", line 269, in run
    self.check_deps(_utils.get_deps_dict(recipe))
  File "/opt/conda/envs/linter/lib/python3.11/site-packages/anaconda_linter/lint/check_build_help.py", line 388, in check_deps
    with open(os.path.join(self.recipe.dir, build_file)) as buildsh:
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen posixpath>", line 90, in join
  File "<frozen genericpath>", line 152, in _check_arg_types
TypeError: join() argument must be str, bytes, or os.PathLike object, not 'CommentedSeq'

ERROR conda.cli.main_run:execute(49): `conda run conda-lint /tmp/abs_29vncl70qt/clone` failed. (See above for error)
</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

INFO: clone/recipe/meta.yaml:0: output_missing_script: output "pre-commit": Output is missing script.
All checks OK

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: clone/recipe/meta.yaml:0: missing_tests: No tests were found.
ERROR: clone/recipe/meta.yaml:1: missing_imports_or_run_test_py: Python packages require imports or a python test file in tests.
ERROR: clone/recipe/meta.yaml:1: missing_pip_check: For pypi packages, pip check should be present in the test commands
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: build_base.bat:0: pip_install_args: output "pre-commit": `pip install` should be run with --no-deps and --no-build-isolation.
ERROR: build_base.sh:2: pip_install_args: output "pre-commit": `pip install` should be run with --no-deps and --no-build-isolation.
ERROR: clone/recipe/meta.yaml:24: outputs_not_unique: output "pre-commit": Output name is not unique
ERROR: clone/recipe/meta.yaml:25: outputs_not_unique: output "pre-commit": Output name is not unique
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

ERROR: build_base.bat:0: pip_install_args: output "pre-commit": `pip install` should be run with --no-deps and --no-build-isolation.
ERROR: build_base.sh:2: pip_install_args: output "pre-commit": `pip install` should be run with --no-deps and --no-build-isolation.
WARNING: clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.
ERROR: clone/recipe/meta.yaml:1: missing_tests: output "pre-commit": No tests were found.
ERROR: clone/recipe/meta.yaml:1: missing_imports_or_run_test_py: output "pre-commit": Python packages require imports or a python test file in tests.
ERROR: clone/recipe/meta.yaml:24: outputs_not_unique: output "pre-commit": Output name is not unique
ERROR: clone/recipe/meta.yaml:25: outputs_not_unique: output "pre-commit": Output name is not unique
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

WARNING: clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.
ERROR: clone/recipe/meta.yaml:1: missing_tests: output "pre-commit": No tests were found.
ERROR: clone/recipe/meta.yaml:1: missing_imports_or_run_test_py: output "pre-commit": Python packages require imports or a python test file in tests.
ERROR: clone/recipe/meta.yaml:24: outputs_not_unique: output "pre-commit": Output name is not unique
ERROR: clone/recipe/meta.yaml:25: outputs_not_unique: output "pre-commit": Output name is not unique
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

WARNING: clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.
ERROR: clone/recipe/meta.yaml:1: missing_tests: output "pre-commit": No tests were found.
ERROR: clone/recipe/meta.yaml:1: missing_imports_or_run_test_py: output "pre-commit": Python packages require imports or a python test file in tests.
ERROR: clone/recipe/meta.yaml:24: outputs_not_unique: output "pre-commit": Output name is not unique
ERROR: clone/recipe/meta.yaml:25: outputs_not_unique: output "pre-commit": Output name is not unique
Errors were found

</details>

anaconda-pkg-build commented:
<details><summary>Linter check found the following problems:</summary>
The following problems have been found:

WARNING: clone/recipe/meta.yaml:0: no_global_test: Global tests are ignored in multi-output recipes.
ERROR: clone/recipe/meta.yaml:1: missing_tests: output "pre-commit": No tests were found.
ERROR: clone/recipe/meta.yaml:1: missing_imports_or_run_test_py: output "pre-commit": Python packages require imports or a python test file in tests.
ERROR: clone/recipe/meta.yaml:26: outputs_not_unique: output "pre-commit": Output name is not unique
ERROR: clone/recipe/meta.yaml:27: outputs_not_unique: output "pre-commit": Output name is not unique
Errors were found

</details>

skupr-anaconda commented:
@JeanChristopheMorinPerso @cbouss Finally, I've fixed the issue on `win-64` by restructuring the recipe. 
My way of thinking was:
1. Build the one-output package without the `outputs` section. It hadn't any errors.
2. Then I renamed the top-level `package/name` from `pre-commit-split` to `pre-commit`. It worked fine but tests skipped for the `pre-commit` output but not for the `pre_commit` one.
3. I moved `tests` from the `pre-commit` output to the top level of the recipe. They passed.

The lesson learned:
- `anaconda-linter` provided incorrect advice for this multi-output recipe, **or** 
- `conda-build` has a bug dealing with multi-output packages.

skupr-anaconda commented:
Staging artifacts are here https://staging.continuum.io/prefect/fs/pre-commit-feedstock/pr5/435da45.
Test it by:
```
conda create -n test
conda activate test
conda install -c "https://staging.continuum.io/prefect/fs/pre-commit-feedstock/pr5/435da45" --repodata-fn repodata.json pre-commit
pre-commit -V
pre-commit --help
```

skupr-anaconda commented:
Rebuilding with `cfgv 3.4.0` https://github.com/AnacondaRecipes/cfgv-feedstock/pull/2

cbouss commented:
> @JeanChristopheMorinPerso @cbouss Finally, I've fixed the issue on `win-64` by restructuring the recipe. My way of thinking was:
> 
> 1. Build the one-output package without the `outputs` section. It hadn't any errors.
> 2. Then I renamed the top-level `package/name` from `pre-commit-split` to `pre-commit`. It worked fine but tests skipped for the `pre-commit` output but not for the `pre_commit` one.
> 3. I moved `tests` from the `pre-commit` output to the top level of the recipe. They passed.
> 
> The lesson learned:
> 
> * `anaconda-linter` provided incorrect advice for this multi-output recipe, **or**
> * `conda-build` has a bug dealing with multi-output packages.

Oh I see what happened now. 
In that case anaconda-linter provides the advice that is best in most cases, and conda-build is just complex.
Had the entry_points been defined on the output level instead of the global level, this would have worked.

skupr-anaconda commented:
> Oh I see what happened now. In that case anaconda-linter provides the advice that is best in most cases, and conda-build is just complex. Had the entry_points been defined on the output level instead of the global level, this would have worked.

I tried to define entry_points on the output level but it's now working at all

JeanChristopheMorinPerso commented:
Looking at the diff, the most notable change is the test section which was moved to the top level of the build recipe. How does that fixes the issue that you were initially trying to solve (ie pre-commit.exe failing to run)?

skupr-anaconda commented:
> Looking at the diff, the most notable change is the test section which was moved to the top level of the build recipe. How does that fixes the issue that you were initially trying to solve (ie pre-commit.exe failing to run)?

My first attempt was to rename the package from `pre-commit-split` to `pre-commit`. I think `conda-build` badly interprets the name of the package and the output one at the same time. It builds fine if the package name **equals** the name of the output. If it's **NOT** equal then entry_points on win will be created incorrectly. That's why I think it's conda-build bug but not the upstream one

JeanChristopheMorinPerso commented:
Thanks for clarifying this. We should create an issue on the conda-build repo if we think there is a bug.

cbouss commented:
By moving entry_points to the output, it does correct the issue. It results in two files:
Scripts\pre-commit.exe and Scripts\pre-commit-script.py

skupr-anaconda commented:
> By moving entry_points to the output, it does correct the issue. It results in two files: Scripts\pre-commit.exe and Scripts\pre-commit-script.py

Let me check again...

skupr-anaconda commented:
> By moving entry_points to the output, it does correct the issue. It results in two files: Scripts\pre-commit.exe and Scripts\pre-commit-script.py

Thanks, Charles. You're right. I **wrongfully** placed entry_points to the output:
```
outputs:
  - name: pre-commit
    script: build_base.bat  # [win]
    script: build_base.sh  # [not win]
    entry_points:
      - pre-commit = pre_commit.main:main
```

The correct way is:
```
outputs:
  - name: pre-commit
    script: build_base.bat  # [win]
    script: build_base.sh  # [not win]
    build:
      entry_points:
        - pre-commit = pre_commit.main:main
```
